overlord/snapstate/reboot.go

```diff
@@ -123,6 +229,241 @@ func bootBaseSnapType(byTypeTss map[snap.Type]*state.TaskSet) snap.Type {
 	return snap.TypeOS
 }
 
+// arrangeSingleRebootForSplitTaskSets sets up restart boundaries for task-sets that have been
+// split into two. It uses the first part of the task-set (i.e the one containing the reboot edges)
+// to set a restart boundary along the 'Do' path for the last task-set, and one along the 'Undo' path
+// for the first task-set, to introduce only one reboot along each direction.
+func arrangeSingleRebootForSplitTaskSets(beforeTss map[snap.Type]*state.TaskSet) error {
+	// Set reboot boundary along the do path for the last essential
+	for i := len(essentialSnapsRestartOrder) - 1; i >= 0; i-- {
+		o := essentialSnapsRestartOrder[i]
+		if beforeTss[o] == nil {
+			continue
+		}
+
+		linkSnap := beforeTss[o].MaybeEdge(EndEdge)
+		if linkSnap == nil {
+			return fmt.Errorf("internal error: no %q edge set in task-set for %q", EndEdge, o)
+		}
+		restart.MarkTaskAsRestartBoundary(linkSnap, restart.RestartBoundaryDirectionDo)
+		break
+	}
+
+	// Set reboot boundary along the undo path for the first essential
+	for _, o := range essentialSnapsRestartOrder {
+		if beforeTss[o] == nil {
+			continue
+		}
+		if unlinkSnap := findUnlinkTask(beforeTss[o]); unlinkSnap != nil {
+			restart.MarkTaskAsRestartBoundary(unlinkSnap, restart.RestartBoundaryDirectionUndo)
+		}
+		break
```

Comment:
"unlink-snap" might exist or not, so maybe the break should be inside the `if` statement?

Reply:
Yea that is more correct, thank you for spotting this.